wcfsetup/install/files/lib/system/service/worker/VAPID.class.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace wcf\system\service\worker;
+
+use Jose\Component\Core\AlgorithmManager;
+use Jose\Component\Signature\Algorithm\ES256;
+use Jose\Component\Signature\JWSBuilder;
+use Jose\Component\Signature\Serializer\CompactSerializer;
+use ParagonIE\ConstantTime\Base64UrlSafe;
+use ParagonIE\ConstantTime\Binary;
+use wcf\data\service\worker\ServiceWorker;
+use wcf\util\JSON;
+
+/**
+ * @author      Olaf Braun
+ * @copyright   2001-2024 WoltLab GmbH
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since       6.1
+ */
+final class VAPID
+{
+    public const PUBLIC_KEY_LENGTH = 65;
+    public const PRIVATE_KEY_LENGTH = 32;
+
+    /**
+     * Add the VAPID header to the request headers.
+     * {@link https://www.rfc-editor.org/rfc/rfc8282}
+     *
+     * @param ServiceWorker $serviceWorker
+     * @param array $headers
+     */
+    public static function addHeader(ServiceWorker $serviceWorker, array &$headers): void
+    {
+        $rawPublicKey = Base64UrlSafe::decode(SERVICE_WORKER_PUBLIC_KEY, true);
+        // Validate the length of the public key
+        if (Binary::safeStrlen($rawPublicKey) !== VAPID::PUBLIC_KEY_LENGTH) {
+            throw new \RuntimeException('Invalid public key length');
+        }
+        ['x' => $x, 'y' => $y] = Util::unserializePublicKey($rawPublicKey);
+
+        $header = [
+            'typ' => 'JWT',
+            'alg' => 'ES256',
+        ];
+        $payload = JSON::encode([
+            'aud' => $serviceWorker->getEndpoint(),
+            'exp' => TIME_NOW + 43200, // 12h
+            'sub' => "mailto:" . MAIL_ADMIN_ADDRESS,
```

Comment:
Should the email address be `rawurlencode`d? Thinking about email addresses such as `admin+myforum@example.com`.

Reply:
This is not needed. The value `mailto:admin+myforum@example.com` works fine.